drivers/frequency/ad9545/ad9545.c

```diff
@@ -0,0 +1,2434 @@
+	}
+	return 0;
+}
+
+static int32_t ad9545_parse_aux_dpll(struct ad9545_dev *dev,
+				     struct ad9545_init_param *init_param)
+{
+	if (!init_param->aux_dpll_clk.dpll_used)
+		return 0;
+
+	dev->aux_dpll_clk.dpll_used = true;
+	dev->aux_dpll_clk.dev = dev;
+	dev->aux_dpll_clk.rate_change_limit =
+		init_param->aux_dpll_clk.rate_change_limit;
+	dev->aux_dpll_clk.source = init_param->aux_dpll_clk.source;
+	dev->aux_dpll_clk.loop_bw_mhz = init_param->aux_dpll_clk.loop_bw_mhz;
+
+	return 0;
+}
+
+static int32_t ad9545_parse_init(struct ad9545_dev *dev,
+				 struct ad9545_init_param *init_param)
+{
+	int32_t ret;
+
+	dev->sys_clk.ref_freq_hz = init_param->sys_clk.ref_freq_hz;
+	dev->sys_clk.sys_clk_crystal = init_param->sys_clk.sys_clk_crystal;
+	dev->sys_clk.sys_clk_freq_doubler = init_param->sys_clk.sys_clk_freq_doubler;
+
+	/* Parse Init Inputs */
+	ret = ad9545_parse_inputs(dev, init_param);
+	if (ret < 0)
+		return ret;
+	/* Parse PLLS */
+	ret = ad9545_parse_plls(dev, init_param);
+	if (ret < 0)
+		return ret;
+	/* Parse Outputs */
+	ret = ad9545_parse_outputs(dev, init_param);
+	if (ret < 0)
+		return ret;
+	/* Parse NCOS*/
+	ret = ad9545_parse_ncos(dev, init_param);
+	if (ret < 0)
+		return ret;
+	/* Parse TDCS*/
+	ret = ad9545_parse_tdcs(dev, init_param);
+	if (ret < 0)
+		return ret;
+	/* Parse DPLL */
+	return ad9545_parse_aux_dpll(dev, init_param);
+}
+
+/**
+ * Initialize the device.
+ * @param device - The device structure.
+ * @param init_param - The structure that contains the device initial
+ *		       parameters.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int32_t ad9545_init(struct ad9545_dev **device,
+		    struct ad9545_init_param *init_param)
+{
+	struct ad9545_dev *dev;
+	int32_t ret;
+
+	dev = (struct ad9545_dev *)no_os_malloc(sizeof(*dev));
+	if (!dev)
+		goto error;
+
+	dev->comm_type = init_param->comm_type;
+	if (dev->comm_type == SPI) {
+		/* SPI */
+		ret = no_os_spi_init(&dev->spi_desc, init_param->spi_init);
+		if (ret < 0)
+			goto error;
+
+		dev->reg_read = ad9545_spi_reg_read;
+		dev->reg_write = ad9545_spi_reg_write;
+		dev->reg_read_multiple = ad9545_spi_reg_read_multiple;
+		dev->reg_write_multiple = ad9545_spi_reg_write_multiple;
+	} else {
+		/* I2C */
+		ret = no_os_i2c_init(&dev->i2c_desc, init_param->i2c_init);
+		if (ret < 0)
+			goto error;
+
+		dev->reg_read = ad9545_i2c_reg_read;
+		dev->reg_write = ad9545_i2c_reg_write;
+		dev->reg_read_multiple = ad9545_i2c_reg_read_multiple;
+		dev->reg_write_multiple = ad9545_i2c_reg_write_multiple;
+	}
+
+	ret = ad9545_write_reg(dev, AD9545_CONFIG_0, AD9545_RESET_REGS);
+	if (ret < 0)
+		return ret;
+
+	if (dev->comm_type == SPI) {
+		/* Enable SDO Line */
+		ret = ad9545_write_reg(dev, AD9545_CONFIG_0, 0x38);
+		if (ret < 0)
+			return ret;
+	}
+
+	/* Check device */
+	ret = ad9545_check_id(dev);
+	if (ret < 0)
+		goto error;
+
+	/* Device settings */
+	ret = ad9545_parse_init(dev, init_param);
+	if (!ret) {
```

Comment:
Looks somehow different than what we have. I'm not saying that it is not correct, only that I am used with the flow below:
if (ret)
   goto error;

*device = dev; ...

Reply:
Sure, i agree with you.